hypervisor/arch/x86/vmx.c

```diff
@@ -360,37 +373,39 @@ void vmx_write_cr0(struct vcpu *vcpu, uint64_t cr0)
 	 * When loading a control register, reserved bit should always set
 	 * to the value previously read.
 	 */
-	cr0 = (cr0 & ~CR0_RESERVED_MASK) |
-		(vcpu_get_cr0(vcpu) & CR0_RESERVED_MASK);
-
-	if (((vcpu_get_efer(vcpu) & MSR_IA32_EFER_LME_BIT) != 0UL) &&
-	    !paging_enabled && ((cr0 & CR0_PG) != 0UL)) {
-		/* Enable long mode */
-		pr_dbg("VMM: Enable long mode");
-		entry_ctrls = exec_vmread32(VMX_ENTRY_CONTROLS);
-		entry_ctrls |= VMX_ENTRY_CTLS_IA32E_MODE;
-		exec_vmwrite32(VMX_ENTRY_CONTROLS, entry_ctrls);
-
-		vcpu_set_efer(vcpu,
-			vcpu_get_efer(vcpu) | MSR_IA32_EFER_LMA_BIT);
-	} else if (((vcpu_get_efer(vcpu) & MSR_IA32_EFER_LME_BIT) != 0UL) &&
-		   paging_enabled && ((cr0 & CR0_PG) == 0UL)){
-		/* Disable long mode */
-		pr_dbg("VMM: Disable long mode");
-		entry_ctrls = exec_vmread32(VMX_ENTRY_CONTROLS);
-		entry_ctrls &= ~VMX_ENTRY_CTLS_IA32E_MODE;
-		exec_vmwrite32(VMX_ENTRY_CONTROLS, entry_ctrls);
-
-		vcpu_set_efer(vcpu,
-			vcpu_get_efer(vcpu) & ~MSR_IA32_EFER_LMA_BIT);
-	} else {
-		/* CR0.PG unchanged. */
+	cr0 &= ~CR0_RESERVED_MASK;
+
+	if (!old_paging_enabled && ((cr0 & CR0_PG) != 0UL)) {
+		if ((vcpu_get_efer(vcpu) & MSR_IA32_EFER_LME_BIT) != 0UL) {
+			/* Enable long mode */
+			pr_dbg("VMM: Enable long mode");
+			entry_ctrls = exec_vmread32(VMX_ENTRY_CONTROLS);
+			entry_ctrls |= VMX_ENTRY_CTLS_IA32E_MODE;
+			exec_vmwrite32(VMX_ENTRY_CONTROLS, entry_ctrls);
+
+			vcpu_set_efer(vcpu,
+				vcpu_get_efer(vcpu) | MSR_IA32_EFER_LMA_BIT);
+		} else if (is_pae(vcpu)) {
+			/* enabled PAE from paging disabled */
+			load_pdptrs(vcpu);
+		}
```

Comment:
we need end this with else {}.   This is also MISRAC requirement

Reply:
done